src/Reopt/TypeInference/Constraints.hs

```diff
@@ -0,0 +1,778 @@
+traceContext :: PP.Doc () -> Context -> Context -> Context
+traceContext description preCtx postCtx =
+  if not traceUnification then postCtx else
+    let msg = PP.vsep [PP.hsep [">>> ", description]
+                      , PP.indent 4 $ PP.pretty preCtx
+                      , PP.hsep ["<<< ", description]
+                      , PP.indent 4 $ PP.pretty postCtx]
+      in trace (show msg) postCtx
+
+traceCOpContext :: PP.Doc () -> TyConstraint -> Context -> Context -> Context
+traceCOpContext fnNm c ctx ctx' =
+  traceContext (fnNm<>"(" <> (PP.pretty $ c)<> ")") ctx ctx'
+
+-- | @solveEqC (s,t) ctx@ updates @ctx@ with the equality
+-- `s == t`. Cf. TIE Algorithm 1. If @Nothing@ is returned, the equality
+-- failed the occurs check. FIXME should we have a `decomposeEqC` similar to `decomposeSubC`?
+solveEqC :: Context -> (Ty,Ty) -> Context
+solveEqC ctx (type1, type2) = traceCOpContext "solveEqC" (EqC type1 type2) ctx $
+  go (substEqs type1 ctx) (substEqs type2 ctx)
+  where go :: Ty -> Ty -> Context
+        go s@(VarTy x) t =
+          if occursIn x t
+            then ctx{ctxOccursCheckFailures=(s,t):(ctxOccursCheckFailures ctx)}
+            else ctx{ctxVarEqMap = Map.insert x t (ctxVarEqMap ctx)}
+        go s t@(VarTy _) = go t s
+        go (PtrTy _w1 s) (PtrTy _w2 t) = ctx {ctxEqConstraints = (s,t):ctxEqConstraints ctx}
+        go s t = if absurdEqC ctx s t
+                 then ctx{ctxAbsurdConstraints = (EqC s t):ctxAbsurdConstraints ctx}
+                 else ctx{ctxDroppedEqConstraints= (s,t):(ctxDroppedEqConstraints ctx)}
+
+
+-- | @upperBoundsClosure s t ctx@ says given `s <: t`, for all type variables
+-- `α` where `α <: s`, we add `α <: t`, update the upper bound of `α`, and
+-- record any implied information from `α <: s`. Cf. rule (2) in the
+-- Decomposition Rules of § 6.3.2. N.B., we interpret the `S <: T` constraint
+-- described by TIE to include any `S` and `T`, but for `∀(α <: S)` to quantify
+-- only over known subtype constraints where `α` is _explicitly a type
+-- variable_.
+upperBoundsClosure :: Ty -> Ty -> Context -> Context
+upperBoundsClosure s t ctx0 = traceCOpContext "upperBoundsClosure" (SubC s t) ctx0 $
+  -- Find all `α` where `α <: s` and perform updates to propogate `α <: t`.
+  Map.foldrWithKey update ctx0 (ctxSubtypeMap ctx0)
+  where
+    update :: TyVar -> (Set Ty, Set Ty) -> Context -> Context
+    update a (lowerBounds,upperBounds) ctx =
+      -- If `α </: s` or if we _already_ know `α <: t` then simply move on.
+      if not (Set.member s upperBounds) || (Set.member t upperBounds) then ctx
+      -- otherwise we need to compute the closure
+      else
+        let -- `S_{<:} = S_{<:} ∪ {a <: t}`
+            supMap' = Map.insert a (lowerBounds, Set.insert t upperBounds) $ ctxSubtypeMap ctx
+            -- `B↑(α) ← B↑(α) ⊓ B↑(T)`
+            boundsMap' = let aLower = fst $ Map.findWithDefault (BotTy, TopTy) a $ ctxVarBoundsMap ctx
```

Comment:
`currentLowerBound` and `currentUpperBound` are perhaps poorly named...? They don't _just_ return the "current bounds", they return the current bound _with no free variables_. I.e., any type variables in the bound are also transitively concretized in the appropriate way, and if we don't actually have an _informative_ bound for them yet, they are replaced by the default bound respectively. We don't want that at this point, we want to leave the type variables there for now. Because of this, I'm fairly confident for that line at least we _don't_ want the transitive lookup, we just want the naive one.

Reply:
I just renamed `currentLowerBound` and `currentUpperBound` to `lowerConcretization` and `upperConcretization` to help make the distinction at least a little clearer.

Reply:
BTW, that reasoning above is definitely a post-hoc justification of why it is the way it is... it could be wrong. I should write a few more tests that rely on variables having related bounds so we're confident nothing wonky is going on...